src/protocols/http/hpr_http_roaming.erl

```diff
@@ -0,0 +1,549 @@
+
+-define(JOIN1_DELAY, 5_000_000).
+-define(JOIN2_DELAY, 6_000_000).
+-define(RX2_DELAY, 2_000_000).
+-define(RX1_DELAY, 1_000_000).
+
+%% Roaming MessageTypes
+-type prstart_req() :: map().
+-type prstart_ans() :: map().
+-type xmitdata_req() :: map().
+-type xmitdata_ans() :: map().
+
+-type netid_num() :: non_neg_integer().
+-type gateway_time() :: non_neg_integer().
+
+-type downlink() :: {
+    ResponseStream :: hpr_router_stream_manager:gateway_stream(),
+    Response :: any()
+}.
+
+-type transaction_id() :: integer().
+-type region() :: atom().
+-type token() :: binary().
+-type dest_url() :: binary().
+-type flow_type() :: sync | async.
+
+-define(TOKEN_SEP, <<"::">>).
+
+-record(packet, {
+    packet_up :: hpr_packet_up:packet(),
+    gateway_time :: gateway_time(),
+    response_stream :: hpr_router_stream_manager:gateway_stream()
+}).
+-type packet() :: #packet{}.
+
+-type downlink_packet() :: hpr_packet_down:packet().
+
+-export_type([
+    netid_num/0,
+    packet/0,
+    gateway_time/0,
+    downlink/0,
+    downlink_packet/0
+]).
+
+%% ------------------------------------------------------------------
+%% Uplink
+%% ------------------------------------------------------------------
+
+-spec new_packet(
+    PacketUp :: hpr_packet_up:packet(),
+    GatewayTime :: gateway_time(),
+    GatewayStream :: hpr_router_stream_manager:gateway_stream()
+) -> #packet{}.
+new_packet(PacketUp, GatewayTime, GatewayStream) ->
+    #packet{
+        packet_up = PacketUp,
+        gateway_time = GatewayTime,
+        response_stream = GatewayStream
+    }.
+
+-spec gateway_time(#packet{}) -> gateway_time().
+gateway_time(#packet{gateway_time = GWTime}) ->
+    GWTime.
+
+-spec response_stream(#packet{}) -> hpr_router_stream_manager:gateway_stream().
+response_stream(#packet{response_stream = ResponseStream}) ->
+    ResponseStream.
+
+-spec make_uplink_payload(
+    NetID :: netid_num(),
+    Uplinks :: list(packet()),
+    TransactionID :: integer(),
+    ProtocolVersion :: pv_1_0 | pv_1_1,
+    DedupWindowSize :: non_neg_integer(),
+    Destination :: binary(),
+    FlowType :: sync | async
+) -> prstart_req().
+make_uplink_payload(
+    NetID,
+    Uplinks,
+    TransactionID,
+    ProtocolVersion,
+    DedupWindowSize,
+    Destination,
+    FlowType
+) ->
+    #packet{
+        packet_up = PacketUp,
+        gateway_time = GatewayTime,
+        response_stream = ResponseStream
+    } = select_best(Uplinks),
+    Payload = hpr_packet_up:payload(PacketUp),
+    PacketTime = hpr_packet_up:timestamp(PacketUp),
+
+    Region = hpr_packet_up:region(PacketUp),
+    DataRate = hpr_packet_up:datarate(PacketUp),
+    Frequency = hpr_packet_up:frequency_mhz(PacketUp),
+
+    {RoutingKey, RoutingValue} = routing_key_and_value(PacketUp),
+
+    Token = make_uplink_token(TransactionID, Region, PacketTime, Destination, FlowType),
+
+    ok = hpr_http_roaming_utils:insert_handler(TransactionID, ResponseStream),
```

Comment:
We can make sure we have some tests for this, but I think we can key the response stream by the pubkeybin now.

This was made when the gateways didn't have persistent connections to the router so we needed to make sure we were replying to the same stream that gave us a packet. That's not longer the case, and I think it means we can simplify how the http protocol stores and looks up handlers.

Reply:
`new_packet` doesn't imply to me that we're changing state. I feel like it would be clearer if the `insert` didn't happen in this function or if the function name implied that we're also remembering something about the packet.

Reply:
hpr_http_roaming_utils:insert_handler and related token functions now take PubKeyBin.